system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,369 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef TOPIC_STATISTICS_COLLECTOR__SUBSCRIBER_TOPIC_STATISTICS_HPP_
+#define TOPIC_STATISTICS_COLLECTOR__SUBSCRIBER_TOPIC_STATISTICS_HPP_
+
+#include <chrono>
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "rcpputils/asserts.hpp"
+
+#include "received_message_age.hpp"
+#include "received_message_period.hpp"
+#include "system_metrics_collector/constants.hpp"
+#include "system_metrics_collector/metrics_message_publisher.hpp"
+#include "topic_statistics_collector.hpp"
+
+using metrics_statistics_msgs::msg::MetricsMessage;
+
+namespace topic_statistics_collector
+{
+
+/**
+ * Build a positive integer range to use in parameter descriptors.
+ *
+ * @return a positive integer range with specififies start, end and step values
+ */
+inline rcl_interfaces::msg::IntegerRange buildPositiveIntegerRange(
+  int64_t from,
+  int64_t to,
+  uint64_t step)
+{
+  rcl_interfaces::msg::IntegerRange positive_range;
+  positive_range.from_value = from;
+  positive_range.to_value = to;
+  positive_range.step = step;
+
+  return positive_range;
+}
+
+/**
+ * Build a parameter description for period node parameters.
+ *
+ * @return a read-only integer range topic descriptor
+ */
+inline rcl_interfaces::msg::ParameterDescriptor buildPeriodParameterDescriptor(
+  const std::string & description)
+{
+  const auto positive_range = buildPositiveIntegerRange(
+    1,
+    std::numeric_limits<decltype(rcl_interfaces::msg::IntegerRange::to_value)>::max(),
+    1
+  );
+
+  rcl_interfaces::msg::ParameterDescriptor period_descriptor;
+  period_descriptor.read_only = true;
+  period_descriptor.integer_range.push_back(positive_range);
+  period_descriptor.description = description;
+
+  return period_descriptor;
+}
+
+/**
+ * Build a parameter description for string node parameters.
+ *
+ * @return a read-only topic descriptor
+ */
+inline rcl_interfaces::msg::ParameterDescriptor buildTopicParameterDescriptor(
+  const std::string & description)
+{
+  rcl_interfaces::msg::ParameterDescriptor topic_descriptor;
+  topic_descriptor.read_only = true;
+
+  return topic_descriptor;
+}
+
+/**
+ * Validate the values assigned to string node parameters.
+ *
+ * @throws std::invalid_argument if the parameter value is empty
+ */
+inline void validateStringParam(const std::string & param)
+{
+  if (param.empty()) {
+    std::stringstream ss;
+    ss << param << " node paramater cannot be empty";
+    throw std::invalid_argument{ss.str().c_str()};
+  }
```

Comment:
```suggestion
  rcpputils::require_true(!param.empty(), "Node parameter cannot be empty.");
```
See https://github.com/ros2/rcpputils/blob/6bc8fa85b6da5aa39a4fc8f5537ca227f1211fe1/include/rcpputils/asserts.hpp#L95

There is also no benefit from including `param` in the exception message since it will be empty.

Reply:
Changed the method to use `rcpputils::require_true` correctly.